prediction/src/algorithms/preprocessing/step1.py

```diff
@@ -0,0 +1,340 @@
+
+        slice_num = int(len(slices) / sec_num)
+        slices.sort(key=lambda x: float(x.InstanceNumber))
+        slices = slices[0:slice_num]
+        slices.sort(key=lambda x: float(x.ImagePositionPatient[2]))
+    try:
+        slice_thickness = np.abs(
+            slices[0].ImagePositionPatient[2] -
+            slices[1].ImagePositionPatient[2])
+    except BaseException:
+        slice_thickness = np.abs(
+            slices[0].SliceLocation - slices[1].SliceLocation)
+
+    for s in slices:
+        s.SliceThickness = slice_thickness
+
+    return slices
+
+
+def get_pixels_hu(slices):
+    image = np.stack([s.pixel_array for s in slices])
+    # Convert to int16 (from sometimes int16),
+    # should be possible as values should always be low enough (<32k)
+    image = image.astype(np.int16)
+
+    # Convert to Hounsfield units (HU)
+    for slice_number in range(len(slices)):
+        intercept = slices[slice_number].RescaleIntercept
+        slope = slices[slice_number].RescaleSlope
+
+        if slope != 1:
+            image[slice_number] = slope * image[slice_number].astype(np.float64)
+            image[slice_number] = image[slice_number].astype(np.int16)
+
+        image[slice_number] += np.int16(intercept)
+
+    return np.array(image, dtype=np.int16), np.array(
+        [slices[0].SliceThickness] + slices[0].PixelSpacing, dtype=np.float32)
+
+
+def binarize_per_slice(image, spacing, intensity_th=-600, sigma=1, area_th=30, eccen_th=0.99, bg_patch_size=10):
+    bw = np.zeros(image.shape, dtype=bool)
+
+    # prepare a mask, with all corner values set to nan
+    image_size = image.shape[1]
+    grid_axis = np.linspace(-image_size / 2 + 0.5, image_size / 2 - 0.5, image_size)
+    x, y = np.meshgrid(grid_axis, grid_axis)
+    d = (x ** 2 + y ** 2) ** 0.5
+    nan_mask = (d < image_size / 2).astype(float)
+    nan_mask[nan_mask == 0] = np.nan
+
+    for i in range(image.shape[0]):
+        # Check if corner pixels are identical, if so the slice  before Gaussian filtering
+        if len(np.unique(image[i, 0:bg_patch_size, 0:bg_patch_size])) == 1:
+            current_bw = scipy.ndimage.filters.gaussian_filter(np.multiply(
+                image[i].astype('float32'), nan_mask), sigma, truncate=2.0) < intensity_th
+        else:
+            current_bw = scipy.ndimage.filters.gaussian_filter(
+                image[i].astype('float32'), sigma, truncate=2.0) < intensity_th
+
+        # select proper components
+        label = measure.label(current_bw)
+        properties = measure.regionprops(label)
+        valid_label = set()
+
+        for prop in properties:
+            if prop.area * spacing[1] * spacing[2] > area_th and prop.eccentricity < eccen_th:
+                valid_label.add(prop.label)
+
+        current_bw = np.in1d(label, list(valid_label)).reshape(label.shape)
+        bw[i] = current_bw
+
+    return bw
+
+
+def all_slice_analysis(bw, spacing, cut_num=0, vol_limit=[0.68, 8.2], area_th=6e3, dist_th=62):
```

Comment:
Could you add a few docstrings so it's easier to grasp what the functions are expecting and doing? :)